logprep/util/rule_loader.py

```diff
@@ -0,0 +1,257 @@
+            case list():
+                for element in self.source:
+                    rule_definitions.extend(
+                        RuleLoader(element, self.processor_name).rule_definitions
+                    )
+        return rule_definitions
+
+    @staticmethod
+    def files_by_extensions(
+        source: str, extensions: Iterable | None = None
+    ) -> Generator[str, None, None]:
+        """return a list of rule files in the source directory."""
+        if extensions is None:
+            extensions = RULE_FILE_EXTENSIONS
+        return (str(path) for path in Path(source).glob("**/*") if path.suffix in extensions)
+
+
+class DirectoryRuleLoader(RuleLoader):
+    """
+    DirectoryRuleLoader is responsible for loading rules from a directory recursively.
+    The directory can contain multiple rule files with supported extensions.
+    In every rule file, multiple rules can be defined.
+    Supported extensions are defined in the RULE_FILE_EXTENSIONS constant.
+
+    Parameters
+    ----------
+    directory : str
+      The path to the directory containing the rule files.
+    Attributes
+    ----------
+    source : str
+      The source directory from which the rule files are loaded.
+    Methods
+    -------
+    rules
+      Returns the list of rules loaded from the directory.
+    """
+
+    def __init__(self, source: str, processor_name: str):
+        self.source = source
+        self.processor_name = processor_name
+
+    @property
+    def rules(self) -> List[Rule]:
+        """
+        Generate and return a list of rules from the specified directory.
+
+        Returns
+        -------
+        list of Rule
+          A list of Rule objects created from the rule files in the directory recursively.
+        """
+        rule_files = self.files_by_extensions(self.source)
+        rule_lists = (FileRuleLoader(str(file), self.processor_name).rules for file in rule_files)
+        return list(itertools.chain(*rule_lists))
+
+    @property
+    def rule_definitions(self) -> List[Dict]:
+        """
+        Generate and return a list of dicts reflecting rule definitions from
+        the specified directory.
+
+        Returns
+        -------
+        list of dict
+          A list of rule definitions created from the rule files in the directory recursively.
+        """
+        rule_files = self.files_by_extensions(self.source)
+        rule_definitions = (
+            FileRuleLoader(str(file), self.processor_name).rule_definitions for file in rule_files
+        )
+        return list(itertools.chain(*rule_definitions))
+
+
+class FileRuleLoader(RuleLoader):
+    """FileRuleLoader is responsible for loading rules from a file.
+    The file can contain multiple rules defined in either JSON or YAML format.
+
+    Parameters
+    ----------
+    source : str
+      The path to the file containing the rule definitions.
+    processor_name : str
+      The name of the processor that the rules belong to.
+    """
+
```

Comment:
Shouln'd there be an `__init__` method like in `DirectoryRuleLoader`, since `source` must be a string?

Reply:
the init is not needed as it is not needed in the other child classes.